File: AccuPay/Domain/Payroll/LeaveAccrualCalculator.vb
Option Strict On

Imports AccuPay.Entity
Imports AccuPay.Tools
Imports AccuPay.Utilities

Public Class LeaveAccrualCalculator

    Public Function Calculate2(employee As Employee,
                               currentDate As Date,
                               leaveAllowance As Decimal,
                               lastTransaction As LeaveTransaction) As Decimal
        Dim nextDate As Date
        If lastTransaction Is Nothing Then
            nextDate = employee.StartDate.AddMonths(1).AddDays(1)
        Else
            nextDate = lastTransaction.TransactionDate.AddMonths(1)
        End If

        ' Check if the current date is still too early to update the ledger
        If currentDate < nextDate Then
            Return 0D
        End If

        Dim leaveHoursPerMonth = leaveAllowance / CalendarConstants.MonthsInAYear

        Dim lastTransactionDate = If(lastTransaction?.TransactionDate, employee.StartDate)

        Dim previousTotalMonths = Calendar.MonthsBetween(employee.StartDate, lastTransactionDate)
        Dim previousTotal = AccuMath.CommercialRound(leaveHoursPerMonth * previousTotalMonths)

        Dim currentTotalMonths = Calendar.MonthsBetween(employee.StartDate, nextDate)
        Dim currentTotal = AccuMath.CommercialRound(leaveHoursPerMonth * currentTotalMonths)

        Dim leaveHours = currentTotal - previousTotal

        Return leaveHours
    End Function

    Public Function Calculate(employee As Employee,
                              currentPayperiod As PayPeriod,
                              leaveAllowance As Decimal,
                              firstPayperiodOfYear As PayPeriod,
                              lastPayperiodOfYear As PayPeriod) As Decimal
        Dim daysInYear = Calendar.DaysBetween(firstPayperiodOfYear.PayFromDate, lastPayperiodOfYear.PayToDate)

        If currentPayperiod.PayFromDate < employee.StartDate Then
            Dim daysInCutoff = Calendar.DaysBetween(employee.StartDate, currentPayperiod.PayToDate)
            Dim leaveHours = ComputeTotalLeave(daysInCutoff, daysInYear, leaveAllowance)

            Return leaveHours
        End If

        If currentPayperiod.RowID = firstPayperiodOfYear.RowID Then
            Dim daysInCutoff = Calendar.DaysBetween(currentPayperiod.PayFromDate, currentPayperiod.PayToDate)
            Dim leaveHours = ComputeTotalLeave(daysInCutoff, daysInYear, leaveAllowance)

            Return leaveHours
        ElseIf currentPayperiod.RowID = lastPayperiodOfYear.RowID Then
            Dim daysInCutoff = Calendar.DaysBetween(currentPayperiod.PayFromDate, currentPayperiod.PayToDate)
            Dim leaveHours = ComputeTotalLeave(daysInCutoff, daysInYear, leaveAllowance)

            Return leaveHours
        Else
            Dim startDateOfYear = firstPayperiodOfYear.PayFromDate

            Dim daysSinceLastCutoff = Calendar.DaysBetween(startDateOfYear, currentPayperiod.PayFromDate.AddDays(-1))
            Dim totalSinceLastCutoff = ComputeTotalLeave(daysSinceLastCutoff, daysInYear, leaveAllowance)

            Dim daysSinceCurrentCutoff = Calendar.DaysBetween(startDateOfYear, currentPayperiod.PayToDate)
            Dim totalSinceCurrentCutoff = ComputeTotalLeave(daysSinceCurrentCutoff, daysInYear, leaveAllowance)

            ' The difference between the total of the last cutoff and the total for the current cutoff will be the
            ' actual leave hours to be given.
            Dim leaveHours = totalSinceCurrentCutoff - totalSinceLastCutoff

            Return leaveHours
        End If
    End Function

    Private Function ComputeTotalLeave(daysPassed As Integer, daysInYear As Integer, leaveAllowance As Decimal) As Decimal
        Dim leaves = (daysPassed / daysInYear) * leaveAllowance

        Return AccuMath.CommercialRound(leaves)
    End Function

End Class

File: Accupay.Data/Services/LeaveAccrualCalculator.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Helpers;
using AccuPay.Utilities;
using System;

namespace AccuPay.Data.Services
{
    public class LeaveAccrualCalculator
    {
        public decimal Calculate2(Employee employee,
                                  DateTime currentAccrualDate,
                                  decimal leaveAllowance,
                                  LeaveTransaction lastAccrual)
        {
            var lastAccrualDate = lastAccrual?.TransactionDate ?? employee.StartDate;
            var leaveHoursPerMonth = decimal.Divide(leaveAllowance, CalendarConstants.MonthsInAYear);

            var previousTotalMonths = Calendar.MonthsBetween(employee.StartDate, lastAccrualDate);
            var previousTotal = AccuMath.CommercialRound(leaveHoursPerMonth * previousTotalMonths);

            var currentTotalMonths = Calendar.MonthsBetween(employee.StartDate, currentAccrualDate);
            var currentTotal = AccuMath.CommercialRound(leaveHoursPerMonth * currentTotalMonths);

            var leaveHours = currentTotal - previousTotal;

            return leaveHours;
        }

        public decimal Calculate(Employee employee,
                                 PayPeriod currentPayperiod,
                                 decimal leaveAllowance,
                                 PayPeriod firstPayperiodOfYear,
                                 PayPeriod lastPayperiodOfYear)
        {
            var daysInYear = Calendar.DaysBetween(firstPayperiodOfYear.PayFromDate, lastPayperiodOfYear.PayToDate);

            if (currentPayperiod.PayFromDate < employee.StartDate)
            {
                var daysInCutoff = Calendar.DaysBetween(employee.StartDate, currentPayperiod.PayToDate);
                var leaveHours = ComputeTotalLeave(daysInCutoff, daysInYear, leaveAllowance);

                return leaveHours;
            }

            if (currentPayperiod.RowID == firstPayperiodOfYear.RowID)
            {
                var daysInCutoff = Calendar.DaysBetween(currentPayperiod.PayFromDate, currentPayperiod.PayToDate);
                var leaveHours = ComputeTotalLeave(daysInCutoff, daysInYear, leaveAllowance);

                return leaveHours;
            }
            // TODO: NEed to implement for the last cutoff
            else if (currentPayperiod.RowID == lastPayperiodOfYear.RowID)
            {
                var daysInCutoff = Calendar.DaysBetween(currentPayperiod.PayFromDate, currentPayperiod.PayToDate);
                var leaveHours = ComputeTotalLeave(daysInCutoff, daysInYear, leaveAllowance);

                return leaveHours;
            }
            else
            {
                var startDateOfYear = firstPayperiodOfYear.PayFromDate;

                var daysSinceLastCutoff = Calendar.DaysBetween(startDateOfYear, currentPayperiod.PayFromDate.AddDays(-1));
                var totalSinceLastCutoff = ComputeTotalLeave(daysSinceLastCutoff, daysInYear, leaveAllowance);

                var daysSinceCurrentCutoff = Calendar.DaysBetween(startDateOfYear, currentPayperiod.PayToDate);
                var totalSinceCurrentCutoff = ComputeTotalLeave(daysSinceCurrentCutoff, daysInYear, leaveAllowance);

                // The difference between the total of the last cutoff and the total for the current cutoff will be the
                // actual leave hours to be given.
                var leaveHours = totalSinceCurrentCutoff - totalSinceLastCutoff;

                return leaveHours;
            }
        }

        private decimal ComputeTotalLeave(int daysPassed, int daysInYear, decimal leaveAllowance)
        {
            var leaves = decimal.Divide(daysPassed, daysInYear) * leaveAllowance;
            //var leaves = (daysPassed / (double)daysInYear) * leaveAllowance;

            return AccuMath.CommercialRound(leaves);
        }
    }
}
